content/en/docs/concepts/extend-kubernetes/compute-storage-net/resource-drivers.md

```diff
@@ -0,0 +1,286 @@
+    // node. If that is not possible, it must return an error. The
+    // controller will call UnsuitableNodes and pass the new information to
+    // the scheduler, which then will lead to selecting a different node
+    // if the current one is not suitable.
+    //
+    // The Claim, ClaimParameters, Class, ClassParameters fields of "claims" parameter
+    // items are read-only and must not be modified. This call must be idempotent.
+    Allocate(ctx context.Context, claims []*ClaimAllocation, selectedNode string)
+
+    // Deallocate gets called when a ResourceClaim is ready to be
+    // freed.
+    //
+    // The claim is read-only and must not be modified. This call must be
+    // idempotent. In particular it must not return an error when the claim
+    // is currently not allocated.
+    //
+    // Deallocate may get called when a previous allocation got
+    // interrupted. Deallocate must then stop any on-going allocation
+    // activity and free resources before returning without an error.
+    Deallocate(ctx context.Context, claim *resourcev1alpha2.ResourceClaim) error
+
+    // UnsuitableNodes checks all pending claims with delayed allocation
+    // for a pod. All claims are ready for allocation by the driver
+    // and parameters have been retrieved.
+    //
+    // The driver may consider each claim in isolation, but it's better
+    // to mark nodes as unsuitable for all claims if it not all claims
+    // can be allocated for it (for example, two GPUs requested but
+    // the node only has one).
+    //
+    // The result of the check is in ClaimAllocation.UnsuitableNodes.
+    // An error indicates that the entire check must be repeated.
+    UnsuitableNodes(ctx context.Context, pod *v1.Pod, claims []*ClaimAllocation, potentialNodes []string) error
+```
+
+
+## Kubelet-plugin implementation
+
+Resource driver's kubelet-plugin's main purpose is to ensure the ResourceClaims, that the Pod will
+be using on Node, have all resources ready for Pod to take in use.
+
+```gRPC
+    // NodePrepareResources prepares several ResourceClaims
+    // for use on the node. If an error is returned, the
+    // response is ignored. Failures for individual claims
+    // can be reported inside NodePrepareResourcesResponse.
+    NodePrepareResources(context.Context, *NodePrepareResourcesRequest) (*NodePrepareResourcesResponse, error)
+    // NodeUnprepareResources is the opposite of NodePrepareResources.
+    // The same error handling rules apply,
+    NodeUnprepareResources(context.Context, *NodeUnprepareResourcesRequest) (*NodeUnprepareResourcesResponse, error)
+```
+
+### Example {#example-pod}
+
+Suppose a Kubernetes cluster is running a resource driver gpu.resource.example.com with Resource
+Class `example.example.com`. Here is an example of a pod requesting this resource to run a demo
+workload:
+
+```yaml
+apiVersion: gpu.resource.example.com/v1alpha1
+kind: GpuClaimParameters
+metadata:
+  name: single-gpu
+spec:
+  count: 1
```

Comment:
Optionally, add a comment that this example depends on a fictional CRD.

Reply:
Done.